hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/expression/json/JsonValueFunction.java

```diff
@@ -0,0 +1,212 @@
+/*
+ * Copyright 2021 Hazelcast Inc.
+ *
+ * Licensed under the Hazelcast Community License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://hazelcast.com/hazelcast-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.sql.impl.expression.json;
+
+import com.google.common.cache.Cache;
+import com.hazelcast.core.HazelcastJsonValue;
+import com.hazelcast.jet.sql.impl.JetSqlSerializerHook;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.expression.VariExpressionWithType;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+import com.jayway.jsonpath.JsonPath;
+import org.apache.calcite.sql.SqlJsonValueEmptyOrErrorBehavior;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.Objects;
+
+import static com.hazelcast.internal.util.StringUtil.isNullOrEmpty;
+
+public class JsonValueFunction<T> extends VariExpressionWithType<T> implements IdentifiedDataSerializable {
```

Comment:
The following query:
```sql
SELECT JSON_VALUE('[1]', CAST(null AS JSON) ERROR ON ERROR)
```
Throws: `IllegalArgumentException: Unexpected SQL type: OTHER`

The query:
```sql
SELECT JSON_VALUE('[1]', null ERROR ON ERROR)
```
Throws: `AssertionError: Operand type is not resolved`

The query:
```sql
SELECT JSON_VALUE('[1]', field1 ERROR ON ERROR) FROM test
```
when `field1` is null throws:
```
java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:880)
	at com.google.common.cache.LocalCache.computeIfAbsent(LocalCache.java:4213)
	at com.hazelcast.jet.sql.impl.expression.json.JsonValueFunction.eval(JsonValueFunction.java:106)
```

Any of the above queries without `ERROR ON ERROR` clause doesn't throw, but they should. In all cases, something like "Null JSON Path expression" is expected.

Same for `JSON_QUERY` (but I didn't test that).

The query:
```sql
SELECT JSON_VALUE('[1]', 'garbage' NULL ON ERROR)
```
doesn't throw, but it should. The `NULL ON ERROR` applies to the invalid json value, not to any other error.